Sources/Subs-Db-mysql.php

```diff
@@ -811,24 +822,82 @@ function smf_db_insert($method = 'replace', $table, $columns, $data, $keys, $ret
 	// Determine the method of insertion.
 	$queryTitle = $method == 'replace' ? 'REPLACE' : ($method == 'ignore' ? 'INSERT IGNORE' : 'INSERT');
 
-	// Do the insert.
-	$smcFunc['db_query']('', '
-		' . $queryTitle . ' INTO ' . $table . '(`' . implode('`, `', $indexed_columns) . '`)
-		VALUES
-			' . implode(',
-			', $insertRows),
-		array(
-			'security_override' => true,
-			'db_error_skip' => $table === $db_prefix . 'log_errors',
-		),
-		$connection
-	);
+	if (!$with_returning || $method != 'ingore')
+	{
+		// Do the insert.
+		$smcFunc['db_query']('', '
+			' . $queryTitle . ' INTO ' . $table . '(`' . implode('`, `', $indexed_columns) . '`)
+			VALUES
+				' . implode(',
+				', $insertRows),
+			array(
+				'security_override' => true,
+				'db_error_skip' => $table === $db_prefix . 'log_errors',
+			),
+			$connection
+		);
+	}
+	else //special way for ignore method with returning
+	{
+		$count = count($insertRows);
+		$ai = 0;
+		for($i = 0; $i < $count; $i++)
+		{
+			$old_id = $smcFunc['db_insert_id']();
+			
+			$smcFunc['db_query']('', '
+				' . $queryTitle . ' INTO ' . $table . '(`' . implode('`, `', $indexed_columns) . '`)
+				VALUES
+					' . $insertRows[$i],
+				array(
+					'security_override' => true,
+					'db_error_skip' => $table === $db_prefix . 'log_errors',
+				),
+				$connection
+			);
+			$new_id = $smcFunc['db_insert_id']();
+			
+			if ($last_id != $new_id) //the inserted value was new
+			{
+				$ai = $new_id;
+			}
+			else	// the inserted value already exists we need to find the pk
+			{
+				$where_string = '';
+				$count2 = count($indexed_columns);
+				for ($x = 0; $x < $count2; $x++)
+				{
+					$where_string += key($indexed_columns[$x]) . ' = '. $insertRows[$i][$x];
+					if (($x + 1) < $count2)
+						$where_string += ' AND ';
+				}
+
+				$request = $smcFunc['db_query']('','
+					SELECT `'. $keys[0] . '` FROM ' . $table .'
+					WHERE ' . $where_string . ' LIMIT 1',
+					array()
+				);
```

Comment:
Well developers should be following the proper logic and making the AI field the first in the list for the keys.  Doesn't mean it will happen 100%, but it could.  I guess we call those bugs in that code.